enterprise/backend/src/metabase_enterprise/serialization/api.clj

```diff
@@ -0,0 +1,237 @@
+(ns metabase-enterprise.serialization.api
+  (:require
+   [clojure.java.io :as io]
+   [compojure.core :refer [POST]]
+   [java-time.api :as t]
+   [metabase-enterprise.serialization.v2.extract :as extract]
+   [metabase-enterprise.serialization.v2.ingest :as v2.ingest]
+   [metabase-enterprise.serialization.v2.load :as v2.load]
+   [metabase-enterprise.serialization.v2.storage :as storage]
+   [metabase.api.common :as api]
+   [metabase.api.routes.common :refer [+auth]]
+   [metabase.logger :as logger]
+   [metabase.models.serialization :as serdes]
+   [metabase.public-settings :as public-settings]
+   [metabase.util :as u]
+   [metabase.util.date-2 :as u.date]
+   [metabase.util.log :as log]
+   [metabase.util.malli.schema :as ms]
+   [ring.core.protocols :as ring.protocols])
+  (:import
+   (java.io File ByteArrayOutputStream)
+   (java.lang AutoCloseable)
+   (org.apache.commons.compress.archivers.tar TarArchiveEntry TarArchiveInputStream TarArchiveOutputStream)
+   (org.apache.commons.compress.compressors.gzip GzipCompressorInputStream GzipCompressorOutputStream GzipParameters)
+   (org.apache.logging.log4j Level)
+   (org.apache.logging.log4j.core.appender AbstractAppender FileAppender)
+   (org.apache.logging.log4j.core.config AbstractConfiguration LoggerConfig)))
+
+(set! *warn-on-reflection* true)
+
+(def ^:dynamic *in-tests* "Set when executed in tests to prevent async removal of files" false)
+
+;;; Request callbacks
+
+(defn- ba-copy [f]
+  (let [baos (ByteArrayOutputStream.)]
+    (io/copy f baos)
+    (.toByteArray baos)))
+
+(defn- on-response! [data callback]
+  (reify
+    ;; Real HTTP requests and mt/user-real-request go here
+    ring.protocols/StreamableResponseBody
+    (write-body-to-stream [_ response out]
+      (ring.protocols/write-body-to-stream data response out)
+      (if *in-tests*
+        (callback)
+        (future (callback))))
+
+    ;; mt/user-http-request goes here
+    clojure.java.io.IOFactory
+    (make-input-stream [_ _]
+      (let [res (io/input-stream (if (instance? File data)
+                                   (ba-copy data)
+                                   data))]
+        (callback)
+        res))))
+
+;;; Storage
+
+(def parent-dir "Dir for storing serialization API export-in-progress and archives."
+  (let [f (io/file (System/getProperty "java.io.tmpdir") "serdesv2-api")]
```

Comment:
`(System/getProperty "java.io.tmpdir")`

does this give a unique temp folder for each java process?

if not, we need to make the name of this folder randomized, otherwise this folder will be shared if we run multiple metabase instances on a single machine.

Reply:
looks like it's platform dependent, https://stackoverflow.com/questions/16352326/when-does-system-getpropertyjava-io-tmpdir-return-c-temp

I think we should make the folder name randomized just to be sure.